checker/src/org/checkerframework/checker/index/searchindex/SearchIndexTransfer.java

```diff
@@ -0,0 +1,106 @@
+import org.checkerframework.dataflow.analysis.TransferInput;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.framework.flow.CFAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFValue;
+
+/**
+ * The transfer function for the SearchIndexFor checker. Allows {@link SearchIndexFor} to be refined
+ * to {@link NegativeIndexFor}.
+ */
+public class SearchIndexTransfer extends IndexAbstractTransfer {
+
+    // The ATF (Annotated Type Factory).
+    private SearchIndexAnnotatedTypeFactory aTypeFactory;
+
+    private CFAnalysis analysis;
+
+    public SearchIndexTransfer(CFAnalysis analysis) {
+        super(analysis);
+        this.analysis = analysis;
+        aTypeFactory = (SearchIndexAnnotatedTypeFactory) analysis.getTypeFactory();
+    }
+
+    /**
+     * {@link org.checkerframework.checker.index.qual.SearchIndexFor {@code @SearchIndexFor}} is
+     * refined only when it is compared to these values.
+     */
+    private enum ValidComparisons {
+        NEGATIVE_ONE,
+        ZERO
+    }
+
+    /**
+     * If the left value is exactly the value of {@code toCompareTo}, and the right side has type
+     * {@link SearchIndexFor}, then the right side's new value in the store should become {@link
+     * NegativeIndexFor}.
+     *
+     * <p>For example, this allows the following code to typecheck:
+     *
+     * <pre>{@code
+     * @SearchIndexFor("a") int index = Arrays.binarySearch(a, y);
+     * if (index < 0) {
+     *     @NegativeIndexFor("a") int negInsertionPoint = index;
+     * }
+     * }</pre>
+     */
+    private void specialHandlingForBinarySearch(
+            Node left, Node right, CFStore store, ValidComparisons toCompareTo) {
+        int valueToCompareTo;
+        switch (toCompareTo) {
+            case ZERO:
+                valueToCompareTo = 0;
+                break;
+            case NEGATIVE_ONE:
+                valueToCompareTo = -1;
+                break;
+            default:
+                // Needed to satisfy Java definite assignment rules.
+                throw new Error("this can't happen");
+        }
+        Long leftValue =
+                IndexUtil.getExactValue(
+                        left.getTree(), aTypeFactory.getValueAnnotatedTypeFactory());
+        if (leftValue != null && leftValue == valueToCompareTo) {
+            AnnotationMirror rightSI =
+                    aTypeFactory.getAnnotationMirror(right.getTree(), SearchIndexFor.class);
+            if (rightSI != null) {
+                List<String> arrays = IndexUtil.getValueOfAnnotationWithStringArgument(rightSI);
+                AnnotationMirror nif = aTypeFactory.createNegativeIndexFor(arrays);
+                store.insertValue(
+                        FlowExpressions.internalReprOf(analysis.getTypeFactory(), right), nif);
+            }
+        }
+    }
+
+    @Override
+    protected void refineGT(
```

Comment:
This is handled by IndexAbstractTransfer, which this class inherits from. It defines < and <= in terms of > and >=, so that only > and >= need to be written. It's also used by the LBC and UBC. I'll make sure that there is a test that demonstrates that this works, though.